File: generators/scales.py
from __future__ import annotations

from music21 import interval, meter, note, scale, stream

from models import NOTE_VALUES
from utils import parse_pitch, validate_range


class IntervalListScale(scale.ConcreteScale):
    """Concrete scale built from semitone offsets relative to tonic."""

    def __init__(self, tonic=None, semitones: list[int] | None = None, **keywords):
        tonic_pitch = parse_pitch(str(tonic)) if tonic is not None else parse_pitch("C4")
        semitone_pattern = semitones or [0, 2, 4, 5, 7, 9, 11, 12]
        pitches = [interval.Interval(s).transposePitch(tonic_pitch) for s in semitone_pattern]
        super().__init__(tonic=tonic_pitch, pitches=pitches, **keywords)


SCALE_FACTORY = {
    "Major": scale.MajorScale,
    "Natural Minor": scale.MinorScale,
    "Harmonic Minor": scale.HarmonicMinorScale,
    "Melodic Minor": scale.MelodicMinorScale,
    "Major Pentatonic": lambda tonic: IntervalListScale(tonic=tonic, semitones=[0, 2, 4, 7, 9, 12]),
    "Minor Pentatonic": lambda tonic: IntervalListScale(tonic=tonic, semitones=[0, 3, 5, 7, 10, 12]),
    "Blues": lambda tonic: IntervalListScale(tonic=tonic, semitones=[0, 3, 5, 6, 7, 10, 12]),
    "Chromatic": scale.ChromaticScale,
}

CYCLE_OF_MAJOR_ROOTS = ["C", "G", "D", "A", "E", "B", "F#", "Db", "Ab", "Eb", "Bb", "F"]


def _find_start_octave(root: str, low: str, high: str, octave_span: int) -> int | None:
    low_p = parse_pitch(low)
    high_p = parse_pitch(high)
    for octave in range(2, 8):
        start = parse_pitch(f"{root}{octave}")
        end = parse_pitch(f"{root}{octave + octave_span}")
        if start.midi >= low_p.midi and end.midi <= high_p.midi:
            return octave
    return None


def _build_scale_instance(scale_name: str, tonic: str):
    factory = SCALE_FACTORY[scale_name]
    return factory(tonic)


def build_scale_exercise(settings: dict) -> tuple[stream.Score | None, str | None, str | None]:
    err = validate_range(settings["lowest_note"], settings["highest_note"])
    if err:
        return None, None, err

    roots = CYCLE_OF_MAJOR_ROOTS if settings.get("all_major_cycle") and settings["scale_type"] == "Major" else [settings["root"]]

    score = stream.Score()
    part = stream.Part()
    part.append(meter.TimeSignature(settings["time_signature"]))

    for root in roots:
        start_octave = _find_start_octave(root, settings["lowest_note"], settings["highest_note"], settings["octave_span"])
        if start_octave is None:
            return None, None, "Selected range is too small for this scale and octave span."

        sc = _build_scale_instance(settings["scale_type"], root)
        start_pitch = parse_pitch(f"{root}{start_octave}")
        end_pitch = parse_pitch(f"{root}{start_octave + settings['octave_span']}")
        asc = sc.getPitches(start_pitch, end_pitch)
        if not asc:
            return None, None, "Could not generate a valid scale with the chosen settings."

        direction = settings["direction"]
        if direction == "Ascending":
            notes = asc
        elif direction == "Descending":
            notes = list(reversed(asc))
        else:
            desc = list(reversed(asc))
            notes = asc + desc[1:]

        for idx, p in enumerate(notes, 1):
            n = note.Note(p)
            n.quarterLength = NOTE_VALUES[settings["note_value"]]
            labels: list[str] = []
            if settings["show_note_names"]:
                labels.append(n.nameWithOctave)
            if settings["show_scale_degrees"]:
                labels.append(str(((idx - 1) % 7) + 1))
            if labels:
                n.lyric = " / ".join(labels)
            part.append(n)

        if len(roots) > 1:
            part.append(note.Rest(quarterLength=1.0))

    score.append(part)
    if len(roots) > 1:
        title = f"All Major Scales - {settings['octave_span']} Octave - {settings['direction']}"
    else:
        title = f"{settings['root']} {settings['scale_type']} Scale - {settings['octave_span']} Octaves - {settings['direction']}"
    return score, title, None

File: src/lib/generators/scales.ts
import { Exercise, GenerateResult, MusicEvent } from "@/lib/music/models";
import { buildScaleDegreeLabels } from "@/lib/music/education";
import { buildScalePitchClasses, parsePitch, scaleKeySignatureLabel } from "@/lib/music/noteUtils";
import { chunkIntoMeasures, uid } from "@/lib/generators/shared";
import { ScaleSettings, validateScaleSettings } from "@/lib/validation/scalesValidation";

const SCALE_INTERVALS: Record<string, number[]> = {
  Major: [0, 2, 4, 5, 7, 9, 11, 12],
  "Natural Minor": [0, 2, 3, 5, 7, 8, 10, 12],
  "Harmonic Minor": [0, 2, 3, 5, 7, 8, 11, 12],
  "Melodic Minor": [0, 2, 3, 5, 7, 9, 11, 12],
  "Major Pentatonic": [0, 2, 4, 7, 9, 12],
  "Minor Pentatonic": [0, 3, 5, 7, 10, 12],
  Blues: [0, 3, 5, 6, 7, 10, 12],
  Chromatic: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
};

const findStartOctave = (root: string, low: string, high: string, octaveSpan: number): number | undefined => {
  const lowMidi = parsePitch(low).midi;
  const highMidi = parsePitch(high).midi;
  for (let octave = 2; octave <= 7; octave += 1) {
    const start = parsePitch(`${root}${octave}`).midi;
    const end = parsePitch(`${root}${octave + octaveSpan}`).midi;
    if (start >= lowMidi && end <= highMidi) return octave;
  }
  return undefined;
};

const buildScaleNotes = (settings: ScaleSettings): { notes?: string[]; error?: string } => {
  const intervals = SCALE_INTERVALS[settings.scaleType];
  if (!intervals) return { error: "Unsupported scale type." };
  const startOct = findStartOctave(settings.root, settings.lowestNote, settings.highestNote, settings.octaveSpan);
  if (startOct === undefined) return { error: "Selected range is too small for this scale." };

  const template = buildScalePitchClasses(settings.root, settings.scaleType, intervals);
  const asc: string[] = [];

  for (let oct = 0; oct < settings.octaveSpan; oct += 1) {
    for (let i = 0; i < template.length - 1; i += 1) {
      const degree = template[i];
      asc.push(`${degree.pitchClass}${startOct + oct + degree.octaveOffset}`);
    }
  }

  const top = template[template.length - 1];
  asc.push(`${top.pitchClass}${startOct + settings.octaveSpan - 1 + top.octaveOffset}`);

  if (settings.direction === "Ascending") return { notes: asc };
  if (settings.direction === "Descending") return { notes: [...asc].reverse() };
  return { notes: [...asc, ...[...asc].reverse().slice(1)] };
};

export const generateScaleExercise = (settings: ScaleSettings): GenerateResult => {
  const err = validateScaleSettings(settings);
  if (err) return { error: err };

  const built = buildScaleNotes(settings);
  if (built.error || !built.notes) return { error: built.error ?? "Unable to generate scale." };

  const degreeLabels = buildScaleDegreeLabels(settings.scaleType, settings.octaveSpan, settings.direction);

  const events: MusicEvent[] = built.notes.map((pitch, idx) => {
    const labels: string[] = [];
    if (settings.showNoteNames) labels.push(pitch);
    if (settings.showScaleDegrees && degreeLabels[idx]) labels.push(degreeLabels[idx]);
    return { kind: "note", pitch, duration: settings.noteValue, lyrics: labels.length ? labels : undefined };
  });

  const exercise: Exercise = {
    id: uid(),
    type: "scale",
    title: `${settings.root} ${settings.scaleType} Scale - ${settings.octaveSpan} Octaves - ${settings.direction}`,
    tempo: "80",
    timeSignature: settings.timeSignature,
    keySignature: scaleKeySignatureLabel(settings.root, settings.scaleType),
    clef: "treble",
    metadata: {
      root: settings.root,
      scaleType: settings.scaleType,
      octaveSpan: settings.octaveSpan,
      direction: settings.direction,
      range: `${settings.lowestNote}-${settings.highestNote}`
    },
    measures: chunkIntoMeasures(events, settings.timeSignature)
  };

  return { exercise };
};
